🤝 ❤️ Add dataset similarity calculation

cthoyt opened the pull request:
This PR is part of a breakdown of #126. It adds a function for calculating the similarity between two re-shuffling of the same dataset. This will support investigation of robustness of a given split.

The first generation of the algorithm calculates the number of triples that need to be moved in and out of the training set, then normalizes by the number of total triples. Effectively, this is a tanimoto similarity of the triples in the training set.

Future directions:

1. A “deterioration” workflow in which triples are randomly removed from the training set. This will enable investigation of how much training data is really needed to perform well, and when you’re removing it, when does performance start to take a hit
2. A “remixing” workflow, which shuffles the training/testing/validation split to enable us to look into robustness of the splits themselves
3. If we want to go outside of modifying a single dataset, we can talk about application specific dataset similarity which could let us look into the construction of the KG itself. For example, a biological KG that has 5 databases vs leaving one of those databases out

cthoyt committed:
Add computation

cthoyt committed:
Add tests

cthoyt committed:
Add dataset level function

Trigger CI

cthoyt committed:
Update base.py

Trigger CI

mberr commented on src/pykeen/datasets/base.py:
```diff
@@ -122,6 +124,22 @@ def get_normalized_name(cls) -> str:
         """Get the normalized name of the dataset."""
         return normalize_string(cls.__name__)
 
+    def similarity(self, other: Dataset) -> float:
```
maybe it would be more extensible to have a method outside of the dataset class itself, which takes two datasets as input and outputs a scalar similarity value?

Right now, it just uses a very specific definition of dataset similarity (jaccard similarity on the sets of triples if I see it correctly).

cthoyt replied:
Well this does just defer to that outside measure. But I guess you’re right - this is pretty limiting.

cthoyt replied:
Also I implemented this really funny way doing 1 - xor/Union rather than directly intersection/Union 😂

cthoyt replied:
The reason was that this gives the intuition of how many swaps you need to make to morph one split into another. Later I will reimplement this to also take into account splits of more than length 2, then the correspondence to tanimoto will go away (so I will leave it as it is)

cthoyt replied:
Addressed in d73a71c3

cthoyt committed:
Merge branch 'master' into add-dataset-similarity

cthoyt committed:
Update interface

Trigger CI

based on @mberr's suggestion, I've exposed the dataset similarity function and made the internal one more extensible for future implementations

cthoyt committed:
Merge branch 'master' into add-dataset-similarity

PyKEEN-bot committed:
Trigger CI
